test/acceptance/test/cluster/cluster.go

```diff
@@ -0,0 +1,482 @@
+)
+
+const charset = "abcdefghijklmnopqrstuvwxyz0123456789"
+
+var seededRand *rand.Rand = rand.New(rand.NewSource(time.Now().UnixNano()))
+
+type KindCluster struct {
+	Name           string
+	Context        string
+	KubeConfigPath string
+}
+
+func NewCluster(name string, context string, kubeConfigPath string) *KindCluster {
+	return &KindCluster{
+		Name:           name,
+		Context:        context,
+		KubeConfigPath: kubeConfigPath,
+	}
+}
+
+func (c *KindCluster) CleanUp() {
+	c.delete()
+	c.deleteKubeConfigFile()
+}
+
+func (c *KindCluster) delete() {
+	cmd := fmt.Sprintf("kind delete cluster --name %s --kubeconfig %s", c.Name, c.KubeConfigPath)
+	command := exec.Command("sh", "-c", cmd)
+	session, err := gexec.Start(command, GinkgoWriter, GinkgoWriter)
+	Expect(err).ShouldNot(HaveOccurred())
+	Eventually(session).Should(gexec.Exit())
+}
+
+func (c *KindCluster) deleteKubeConfigFile() {
+	err := os.RemoveAll(c.KubeConfigPath)
+	Expect(err).ShouldNot(HaveOccurred())
+}
+
+func CreateKindCluster(ctx context.Context, rootKubeConfigFilesPath string) (*KindCluster, error) {
+	supportedK8SVersions := "1.19.1, 1.20.2, 1.21.1"
+
+	k8sVersion, found := os.LookupEnv("K8S_VERSION")
+	if !found {
+		k8sVersion = "1.20.2"
+	}
+
+	if !strings.Contains(supportedK8SVersions, k8sVersion) {
+		log.Errorf("Kubernetes version %s is not supported for testing", k8sVersion)
+		return nil, errors.New("unsupported kubernetes version")
+	}
+
+	var cluster *KindCluster
+
+	clusterName := RandString(30)
+	kubeConfigFile := "kube-config-" + clusterName
+	kubeConfigPath := filepath.Join(rootKubeConfigFilesPath, kubeConfigFile)
+
+	log.Infof("Creating a kind cluster %s", clusterName)
+
+	c := fmt.Sprintf("kind create cluster --name=%s --kubeconfig %s --image=%s --config=configs/kind-config.yaml --wait 5m", clusterName, kubeConfigPath, "kindest/node:v"+k8sVersion)
+	cmd := exec.CommandContext(ctx, "sh", "-c", c)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	err := cmd.Run()
+	if err != nil {
+		log.Info("Failed to create kind cluster")
+		log.Fatal(err)
+
+		return nil, err
+	}
+
+	cluster = NewCluster(clusterName, fmt.Sprintf("kind-%s", clusterName), kubeConfigPath)
+
+	return cluster, nil
+}
+
+func StringWithCharset(length int, charset string) string {
+	b := make([]byte, length)
+	for i := range b {
+		b[i] = charset[seededRand.Intn(len(charset))]
+	}
+
+	return string(b)
+}
+
+func RandString(length int) string {
+	return StringWithCharset(length, charset)
+}
+
+type ClusterStatus string
+
+const (
+	ClusterCreating  ClusterStatus = "CREATING"
+	ClusterRequested ClusterStatus = "REQUESTED"
+	ClusterCreated   ClusterStatus = "CREATED"
+	ClusterBeingUsed ClusterStatus = "BEINGUSED"
+	ClusterDeleted   ClusterStatus = "DELETED"
+)
+
+const CLUSTER_DB = "db"
+const CLUSTER_TABLE = "clusters"
+
+type Cluster2 struct {
+	Name            string
+	Context         string
+	KubeConfigPath  string
+	Status          ClusterStatus
+	ErrorOnCreation error
+}
```

Comment:
I am not sure i can see why we have cluster2. Could this be included with KindCluster. Basically I am wondering why we have functions to create 2 of everything? Couldnt just one function handle it and have var's passed to handle different clusters being created?

Reply:
Same question. I expected `Cluster2` to be a typo.

Reply:
My bad, after many attempts to implement this, I didn't realize this could be improved. I did a unification of `KindCluster` and `Cluster2` into just `Cluster`. Good catch guys thank you. 🙂